src/Operation/BulkWrite.php

```diff
@@ -341,28 +341,30 @@ public function execute(Server $server)
                     break;
 
                 case self::INSERT_ONE:
-                    $insertedDocument = isset($this->options['codec'])
-                        ? $this->options['codec']->encodeIfSupported($args[0])
-                        : $args[0];
-                    // Psalm's assert-if-true annotation does not work with unions, so
-                    // assert the type manually instead of using is_document
-                    // See https://github.com/vimeo/psalm/issues/6831
-                    assert(is_array($insertedDocument) || is_object($insertedDocument));
-
-                    $insertedIds[$i] = $bulk->insert($insertedDocument);
+                    if (isset($this->options['codec'])) {
+                        $args[0] = $this->options['codec']->encodeIfSupported($args[0]);
+
+                        // Psalm's assert-if-true annotation does not work with unions, so
+                        // assert the type manually instead of using is_document
+                        // See https://github.com/vimeo/psalm/issues/6831
+                        assert(is_array($args[0]) || is_object($args[0]));
+                    }
+
+                    $insertedIds[$i] = $bulk->insert($args[0]);
                     break;
 
                 case self::REPLACE_ONE:
-                    $replacementDocument = isset($this->options['codec'])
-                        ? $this->options['codec']->encodeIfSupported($args[1])
-                        : $args[1];
-                    // Psalm's assert-if-true annotation does not work with unions, so
-                    // assert the type manually instead of using is_document
-                    // See https://github.com/vimeo/psalm/issues/6831
-                    assert(is_array($replacementDocument) || is_object($replacementDocument));
-
-                    $bulk->update($args[0], $replacementDocument, $args[2]);
-                    break;
+                    if (isset($this->options['codec'])) {
+                        $args[1] = $this->options['codec']->encodeIfSupported($args[1]);
+
+                        // Psalm's assert-if-true annotation does not work with unions, so
+                        // assert the type manually instead of using is_document
+                        // See https://github.com/vimeo/psalm/issues/6831
+                        assert(is_array($args[1]) || is_object($args[1]));
```

Comment:
Similar to `INSERT_ONE` above, I wonder if we should duplicate the replacement document validation here.

The constructor performs additional checks on `$args[1]` that we're missing here:

 * Treat empty arrays as replacement documents for BC
 * Prohibit `is_first_key_operator($args[1])`
 * Prohibit `is_pipeline($args[1], true /* allowEmpty */)`

----

Having said this, I'm now wondering why you decided to encode at execution time instead of assembly time in the constructor:

```php
foreach ($operations as $i => $operation) {
    ...

    switch ($type) {
        case self::INSERT_ONE:
            // here

        ...

        case self::REPLACE_ONE:
            // and here
```

I realize `encodeIfSupported()` doesn't throw (at least our implementations shouldn't), but this would consolidate all of the validation errors to construction time. If we're going to be utilizing a codec, I think it makes sense to _only_ validate the result of the encoding instead of both sides as we seem to be doing here (and noting that the post-encoding validation is incomplete).